doc/crdt-details.md

```diff
@@ -0,0 +1,1054 @@
+
+The idea behind how it works is that we already have all the characters we need in the `text` and `tombstones` `Rope`s we store, but some of the characters from the past revision might have been deleted, and some new characters might have been inserted that weren't in the past revision. We need to find a way to delete the newer insertions from `text` and insert the things that weren't deleted at the past point from where they are in `tombstones`.
+
+The way we describe the current state of `text` and `tombstones` relative to the "union string" is with `deletes_from_union` (see [Engine](#engine)), so what if we could find a similar `old_deletes_from_cur_union` that represented what the old revision's text looked like relative to the current union string. This would be the same as our current `deletes_from_union` except characters inserted after the old revision would be marked deleted and newer deletes would be un-marked. The function that finds this is `Engine::deletes_from_cur_union_for_index`.
+
+Once we have this `old_deletes_from_cur_union` and a new `deletes_from_union`, we need a way to take our current `text` and `tombstones` and get a `Rope` of what the `text` would have looked like at that old revision. We can do this by performing inserts and deletes on the `text` `Rope` based on the differences between the old and new deletions. We already have a way of describing inserts and deletes (a `Delta`), and we can create one using a helper called `Delta::synthesize`.
+
+Then we just have to apply the `Delta` we synthesized to the current `text`, returning the resulting old text.
+
+```rust
+/// Get text of a given revision, if it can be found.
+pub fn get_rev(&self, rev: RevToken) -> Option<Rope> {
+    self.find_rev_token(rev).map(|rev_index| self.rev_content_for_index(rev_index))
+}
+
+/// Get text of a given revision, if it can be found.
+fn rev_content_for_index(&self, rev_index: usize) -> Rope {
+    let old_deletes_from_union = self.deletes_from_cur_union_for_index(rev_index);
+    let delta = Delta::synthesize(&self.tombstones,
+        &self.deletes_from_union, &old_deletes_from_union);
+    delta.apply(&self.text)
+}
+```
+
+#### Engine::deletes_from_cur_union_for_index
+
+If you look back at the [example history scenario](#example-history), you'll see `back_computed_deletions_from_6_union`, which shows that for any past `Revision` we can find a set of deletions from the current union string that result in the past text. This helper is what computes deletion sets like `back_computed_deletions_from_6_union`.
+
+We can find an `old_deletes_from_cur_union` by taking our current `deletes_from_union` and walking backwards through our list of `Revision`s undoing the changes they have made since the old revision.
+
+There's a problem with this though, the `inserts` and `deletes` subsets use indices in the coordinate space of the union string at the time the `Revision` was created, which may be smaller than our current union string.
```

Comment:
There's a problem with this though, the `inserts` and `deletes` subsets **at a given `Revision`** use indices in the coordinate